### PR TITLE
brotli support in gzip plugin

### PR DESCRIPTION
**Pull request by myraid**
The requirements are outlined in
https://github.com/apache/trafficserver/issues/1555

For brotli to work the "proxy.config.http.normalize_ae_gzip" configuration should be set to '0'
Should we control this flag manually or through the gzip plugin?

@bryancall @shukitchan can you review.

### CODE REVIEW

**Comment by myraid** [comment]
For brotli to work the "proxy.config.http.normalize_ae_gzip" configuration should be set to '0'
Should we control this flag manually or throught the gzip plugin?

**Comment by bryancall** [comment]
Added the requirements here:
```
Here are the requirements for supporting brotli encoding.

If the Accept-Encoding has brotli, then the normalize AE should pick up brotli. (Should we send both "br" and "gzip" in AE?
If the response is already encoded then do nothing. Should we have a configuration to force brotli encode the response?
If the response is unencoded and if Accept-Encoding has brotli then encode using brotli.
```

**Comment by maskit** [comment]
Having one plugin for multiple compression algorithms is fine with me. However, I'm not a big fun of adding it with "else if" statements, though we already did that for deflate. It looks like ifdef things and it's not so easy to read.

I think it should have an abstract parent class for all the algorithms and each algorithms should be implemented as subclasses.

**Comment by shukitchan** [comment]
I will take a closer look in the coming days.
In the meanwhile, can you update the documentation for the gzip plugin as well for this change?

**Comment by zwoop** [comment]
I need to read this some more, but it feels like we need a configuration option here as well.

**Comment by myraid** [comment]
@bryancall @zwoop  I am working on a enhancement to force brotli compression even if content-encoding header is present.
Do you think it will be a good idea to have a separate plugin for unzip and combine it with this plugin if there is a need for force encoding brotli?

**Comment by bryancall** [comment]
It looks like brotli is on by default?  It would be nice to have an option on what types of compression algorithms that you want to enable.

File: plugins/gzip/gzip.cc [hunk 1]
```diff
@@ -95,18 +104,29 @@ gzip_data_alloc(int compression_type)
     }
   }
 
+  if(compression_type == COMPRESSION_TYPE_BROTLI) {
+  	debug("gzip-transform: brotli compression. Create Brotli Encoder Instance.");
```
Comment (myraid) [inline plugins/gzip/gzip.cc]: Done.

**Comment by myraid** [comment]
@bryancall Yes the default behavior is br. Are you suggesting that the option should dictate normalizing the AE? What will be the behavior when both gzip and br are selected?

**Comment by bryancall** [comment]
@myraid 
I can see cases where someone might not want to spend the CPU resources on Brotli compression, but doesn't mind if the origin does Broli encoding.  The plugin would compress the response from the origin using certain algorithms.

The configuration option would be something like:
algorithms deflate,gzip,brotli

It would really be nice to enable Brotli on certain content types (e.g. text/javascript) that are static and not on types that would be more dynamic (e.g. text/html).  Unfortunately the configuration doesn't easily lend itself to this.

File: plugins/gzip/gzip.cc [hunk 2]
```diff
@@ -350,30 +435,89 @@ gzip_transform_finish(GzipData *data)
...
+      }
+
+      break;
+    }
+
+    if (data->downstream_length != (int64_t)(data->bstrm.total_out)) {
+      error("brotli-transform: ERROR: output lengths don't match (%d, %ld)", data->downstream_length, data->bstrm.total_out);
+    }
+    debug("brotli-transform: Finished brotli");
+    // gzip_log_ratio(data->bstrm.total_in, data->downstream_length);
```
Comment (shukitchan) [inline plugins/gzip/gzip.cc]: Can we still do similar logging brotli?

**Comment by shukitchan** [comment]
My feedbacks

1) Perhaps we are overloading the "Accept-Encoding" header in the request. Let's say we create another header called "X-Normalized-Accept-Encoding". Its value will be the best algorithm we allow (through the configuration option) in the plugin chosen from the incoming "Accept-Encoding" of the request. 

2) We need to change the incoming "Vary" header from response (in read_response hook) and replace "Accept-Encoding" with "X-Normalized-Accept-Encoding". That will cause the cache to only cache according to X-Normalized-Accept-Encoding instead of "Accept-Encoding". 

3) Now we only force brotli encryption or gzip encryption (whichever X-Normalized-Accept-Encoding indicates) when the origin returns plain text.  

4) If we want to force brotli encryption for certain url patterns, we probably can use lua/header_write plugin to change the incoming "Accept-Encoding" to remove "gzip" and "deflate" and leave "br" value in before the gzip plugin runs. This feature can be baked into the gzip plugin later on with more configuration options

The pros is that 
1) We send the "Accept-Encoding" over to the origin and let origin decides what compression to use. 
2) We still only caches a limited amount of copy per URL. The worst case is one per compression algorithm we allow from the plugin configuration option. 

The cons is that 
1) We should remove X-Normalized-Accept-Encoding in send_request hook.
2) We should replace X-Normalized-Accept-Encoding with Accept-Encoding to the Vary response header in send_response hook. 

Any thoughts? @zwoop @bryancall @myriad ?

**Comment by zwoop** [comment]
See also ##776.

**Comment by myraid** [comment]
Added the supported-algorithms cofiguration. 
Normalizing Accept-Encoding now supports both br and gzip together.  (Accept-Encoding: br, gzip)
If the origin supports brotli compression then the plugin just sends the data downstream.
If origin does not support brotli, then origin can still fallback to gzip. 

compression will be triggered for  response based on Accept-Encoding and supported algorithms (brotli is checked before gzip). A transform is not even created if the there is no intersect between Accept-Encoding and supported algorithms.

Caching scenarios.

We only cache 2 copies 

- with 'Accept-Encoding: br, gzip'  -- if content encoding header is present, just cache it. Otherwise encode the content and cache. The User-Agent might receive either br or gzip encoded response which is OK.

- with 'Accept-Encoding: gzip'  -- The User-Agent will always receive a gzip encoded response.

**Comment by shukitchan** [comment]
@myraid your description seems to be good. i will spend some times tomorrow to review the code. 

meanwhile, pls update this as well 
https://github.com/apache/trafficserver/blob/master/doc/admin-guide/plugins/gzip.en.rst
and perhaps this as well 
https://github.com/apache/trafficserver/blob/master/plugins/gzip/README

**Comment by shukitchan** [comment]
two more points 

1) I think we need to maintain the existing behavior as much as possible. So perhaps the default for the new configuration option "supported-algorithms" should not be a no-op? Also it would be great to update the sample configuration file to show how the new option can be used.

2) One problem with "Accept-Encoding: br, gzip" is that the origin will likely just return the gzip version almost all the time. And therefore we can't take advantage of br in almost all cases. Can we perhaps introduce an additional configuration option called "enforce-brotli" ? e.g. If I have "enforce-brotli /notthis/*.js" in the config file, then I will set normalize "Accept-Encoding" to "br" instead of "br, gzip" or "gzip" for the matched URL path. 

P.S. I think enforce-brotli and supported-algorithms should be independent of each other.

**Comment by myraid** [comment]
@shukitchan  My Comments inline

1. I think we need to maintain the existing behavior as much as possible. So perhaps the default for the new configuration option "supported-algorithms" should not be a no-op? Also it would be great to update the sample configuration file to show how the new option can be used.

> [myraid] I agree.  the "supported-algorithms" configuration will be honored if present, otherwise compression could be defaulted to gzip.

2. One problem with "Accept-Encoding: br, gzip" is that the origin will likely just return the gzip version almost all the time. And therefore we can't take advantage of br in almost all cases. Can we perhaps introduce an additional configuration option called "enforce-brotli" ? e.g. If I have "enforce-brotli /notthis/*.js" in the config file, then I will set normalize "Accept-Encoding" to "br" instead of "br, gzip" or "gzip" for the matched URL path.

> [myraid] There is a configuration option "hide-accept-encoding" and if present will hide the Accept-Encoding header. Using this flag we can easily trick the origin to send unencoded data. 
I agree that we should have the "enfore-brotli" flag to provide a config option  to always enforce brotli compression even if content-encoding is present. e.g. If the origin sends gzip, then the plugin should still be able to inflate  and compress back using brotli. I am working on introducing this option.

**Comment by shukitchan** [comment]
wouldn't it be easier if "enforce-brotli" will override the "Accept-Encoding" to "br"? 
So the origin will either return plain text or brotli encoded data. And the plugin only need to brotli-encode the plain text version if that's returned. 

Another concern @bryancall made in an email with us is whether we should only do this brotli encoding transform only if response is cacheable. I wonder whether we should do it in this PR or have a custom version internally instead.

**Comment by myraid** [comment]
Yes we could enforce brotli encoding by overriding the Accept-Encoding to "br" if we are OK with origin responding with plain text. Infact this is cleaner approach. 
 Do we have to worry about the use case where origin sends gzip and the plugin needs to compress it to brotli? Most origins support gzip and is a huge bandwidth saver. 

May be you are thinking of a separate unzip plugin. Ether ways I am fine. Infact we can keep this plugin clean without adding gunzip functionality.

@shukitchan I will work the changes  you suggested.

**Comment by myraid** [comment]
@shukitchan @bryancall can you guys review.

**Comment by myraid** [comment]
@shukitchan @bryancall can you review the new code.
changed enforce_brotli to wildcard pattern matching for urls. 
Now enforce_brotli will override some of the existing configuration for url that matched the pattern.

File: plugins/gzip/gzip.cc [hunk 3]
```diff
@@ -115,20 +142,32 @@ gzip_data_destroy(GzipData *data)
...
-gzip_content_encoding_header(TSMBuffer bufp, TSMLoc hdr_loc, const int compression_type)
+content_encoding_header(TSMBuffer bufp, TSMLoc hdr_loc, const int compression_type, int algorithm)
 {
   TSReturnCode ret;
   TSMLoc ce_loc;
-
+  const char *value = nullptr;
+  int value_len     = 0;
   // Delete Content-Encoding if present???
+  if (compression_type & COMPRESSION_TYPE_BROTLI && (algorithm & ALGORITHM_BROTLI)) {
```
Comment (shukitchan) [inline plugins/gzip/gzip.cc]: so does that mean i must have "br" in supported-algorithm for the content header to be set accordingly?

**Comment by shukitchan** [comment]
[approve ci]

**Comment by shukitchan** [comment]
it looks good to me. 
@bryancall @zwoop can either of you guys take a look, too? thanks.

**Comment by atsci** [comment]
AU check *successful*! https://ci.trafficserver.apache.org/job/autest-github/91/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/94/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/107/

**Comment by atsci** [comment]
Linux build *failed*! https://ci.trafficserver.apache.org/job/linux-github/1682/

**Comment by atsci** [comment]
FreeBSD11 build *failed*! https://ci.trafficserver.apache.org/job/freebsd-github/1789/

**Comment by atsci** [comment]
Intel CC build *failed*! https://ci.trafficserver.apache.org/job/icc-github/220/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/352/

**Comment by myraid** [comment]
fixed the build failures with preprocessor derivatives.
@shukitchan @bryancall @zwoop is this approach OK?

**Comment by shukitchan** [comment]
[approve ci]

**Comment by atsci** [comment]
AU check *successful*! https://ci.trafficserver.apache.org/job/autest-github/95/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/98/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/112/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/224/

**Comment by atsci** [comment]
FreeBSD11 build *failed*! https://ci.trafficserver.apache.org/job/freebsd-github/1794/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/1686/

**Comment by shukitchan** [comment]
please check the failure.

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/356/

**Comment by shukitchan** [comment]
[approve ci]

**Comment by atsci** [comment]
AU check *successful*! https://ci.trafficserver.apache.org/job/autest-github/96/

**Comment by atsci** [comment]
clang format *failed*! https://ci.trafficserver.apache.org/job/clang-format-github/99/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/113/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/225/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1795/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/1687/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/357/

**Comment by shukitchan** [comment]
[approve ci]

**Comment by atsci** [comment]
AU check *successful*! https://ci.trafficserver.apache.org/job/autest-github/97/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/100/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/114/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1796/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/1688/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/226/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/358/

File: doc/admin-guide/plugins/gzip.en.rst [hunk 4]
```diff
@@ -162,6 +182,32 @@ might create a configuration with the following options::
     disallow /notthis/*.js
     flush true
 
+    # Allows brotli encoded response from origin but is not capable of brotli compression
+    [brotli.allowed.com]
+    enabled true
+    compressible-content-type text/*
+    compressible-content-type application/json
+    flush true
+    supported-algothims gzip,deflate
```
Comment (bryancall) [inline doc/admin-guide/plugins/gzip.en.rst]: spelling mistake "supported-algothims" should be "supported-algorithms"

File: doc/admin-guide/plugins/gzip.en.rst [hunk 5]
```diff
@@ -162,6 +182,32 @@ might create a configuration with the following options::
...
+
+    # Enforces brotli compression for javascript and css files
+    [brotli.enforce.com]
+    enabled true
+    compressible-content-type text/*
+    compressible-content-type application/json
+    flush true
+    supported-algothims gzip
```
Comment (bryancall) [inline doc/admin-guide/plugins/gzip.en.rst]: spelling mistake "supported-algothims" should be "supported-algorithms"

File: doc/admin-guide/plugins/gzip.en.rst [hunk 6]
```diff
@@ -162,6 +182,32 @@ might create a configuration with the following options::
...
+    enforce-brotli /this/*.js
+    enforce-brotli /this/*.css
+
+    # Supports brotli compression
+    [brotli.enforce.com]
+    enabled true
+    compressible-content-type text/*
+    compressible-content-type application/json
+    flush true
+    supported-algothims br, gzip
```
Comment (bryancall) [inline doc/admin-guide/plugins/gzip.en.rst]: spelling mistake "supported-algothims" should be "supported-algorithms"

**Comment by bryancall** [comment]
Wouldn't it better to have an configuration such as:
```
compressible-path /this/*.js
compressible-path /this/*.css
# with 
supported-algothims br

vs
enforce-brotli /this/*.js
enforce-brotli /this/*.css

I am not a big fan of having algorithm specific configuration options.
```

**Comment by bryancall** [comment]
I am testing out the plugin with this configuration and I am not getting to to compress using Brotli:

```
[bcall@C02R70EHG8WP-lm ~]$ tail -1 /usr/local/etc/trafficserver/plugin.config
gzip.so /usr/local/etc/trafficserver/gzip.config

[bcall@C02R70EHG8WP-lm ~]$ tail -1 /usr/local/etc/trafficserver/remap.config
map / https://s.yimg.com/

[bcall@C02R70EHG8WP-lm ~]$ cat /usr/local/etc/trafficserver/gzip.config
# Enforces brotli compression for javascript and css files
[www.yahoo.com]
remove-accept-encoding true
enabled true
flush true
supported-algorithms bz
compressible-content-type text/*
enforce-brotli *.js
enforce-brotli *.css

[bcall@C02R70EHG8WP-lm ~]$ curl -H 'Host: s.yimg.com' -H 'accept-encoding: br' -s -o /dev/null -D - 'http://127.0.0.1:8080/zz/combo?/os/stencil/3.1.0/styles-ltr.css&/os/yc/css/bundle.c60a6d54.css'
HTTP/1.1 200 OK
Vary: Accept-Encoding
Last-Modified: Wed, 08 Feb 2017 01:24:18 GMT
Content-Type: text/css
Expires: Fri, 03 Feb 2034 01:51:59 GMT
Cache-Control: max-age=536112000, public
Date: Wed, 08 Feb 2017 01:24:18 GMT
Age: 4138271
Content-Length: 115395
Via: http/1.0 c1.ycs.bf1.yahoo.com (ApacheTrafficServer [cRs f ]), https/1.1 e19.ycpi.cha.yahoo.com (ApacheTrafficServer [cRs f ])
Server: ATS/7.2.0
Connection: keep-alive
```

**Comment by myraid** [comment]
@bryancall You will need brotli libraries in your build root. Did you install the libraries?

**Comment by bryancall** [comment]
@myraid There was an issue with my build.  I got it working.  It would would be great to have debug and error warnings if you haven't built it with brotli and have it configured that way.  I am sure others will run into this issue.

**Comment by bryancall** [comment]
Here are the full debug messages without brotli compiled with the plugin and there are no error messages.  I get the same debug/error messages when I also configure the "asdf" compression algorithm.

```
[bcall@C02R70EHG8WP-lm trafficserver]$ sudo /usr/local/bin/traffic_server -T gzip
traffic_server: using root directory '/usr/local'
[Mar 28 09:55:23.660] Server {0x7fffcca323c0} DIAG: (gzip) INFO: TSPluginInit gzip.so
[Mar 28 09:55:23.660] Server {0x7fffcca323c0} DIAG: (gzip) INFO: Parsing file "/usr/local/etc/trafficserver/gzip.config"
[Mar 28 09:55:23.660] Server {0x7fffcca323c0} DIAG: (gzip) DEBUG: [gzip/gzip.cc:987] [load_global_configuration] config swapped, old config 0x0
[Mar 28 09:55:23.660] Server {0x7fffcca323c0} DIAG: (gzip) INFO: loaded
[Mar 28 09:55:34.045] Server {0xb0016000} DIAG: (gzip) INFO: Kicking off gzip plugin for request
[Mar 28 09:55:34.045] Server {0xb0016000} DIAG: (gzip) INFO: normalized accept encoding to br
[Mar 28 09:55:34.045] Server {0xb0016000} DIAG: (gzip) INFO: handling compression of cached object
[Mar 28 09:55:34.045] Server {0xb0016000} DIAG: (gzip) INFO: No compression algorithms configured
```

**Comment by myraid** [comment]
@bryancall added error messages.

> enforce-brotli /this/*.js

is introduced to selectively turn on  brotli compression for a particular path with minimal changes to the existing deployment.

**Comment by bryancall** [comment]
We should rip out enforce-brotli configuration option the same can be done with multiple configs like the example below.  We might want to add an allow feature and use the ! syntax like other options. but we can add that later.

```
[s.yimg.com]
remove-accept-encoding true
enabled true
flush true
compressible-content-type text/*
disallow *.js
disallow *.css
supported-algorithms gzip

[s.yimg.com]
remove-accept-encoding true
enabled true
flush true
compressible-content-type text/*
supported-algorithms br
```

File: doc/admin-guide/plugins/gzip.en.rst [hunk 7]
```diff
@@ -141,6 +141,26 @@ will leave the header intact if the client provided it.
...
 
+supported-algorithms
+----------------------
+
+Provides the compression algorithms that are supported. This will allow the proxy to selectively
+support certain compressions. The default is gzip. Multiple algorthims can be selected using ',' delimiter
+
+-- To selectively support only certain compression algorithms.
+
+enforce-brotli
```
Comment (bryancall) [inline doc/admin-guide/plugins/gzip.en.rst]: We should rip out enforce-brotli configuration option the same can be done with multiple configs like the example below.  We might want to add an allow feature and use the ! syntax like other options, but we can add that later.

```
[s.yimg.com]
remove-accept-encoding true
enabled true
flush true
compressible-content-type text/*
disallow *.js
disallow *.css
supported-algorithms gzip

[s.yimg.com]
remove-accept-encoding true
enabled true
flush true
compressible-content-type text/*
supported-algorithms br
```

Comment (bryancall) [inline doc/admin-guide/plugins/gzip.en.rst]: Also, please squash the commits

**Comment by myraid** [comment]
@bryancall I removed the enforce-brotli config option. Squashed all the commits.

Can you review and approve.

**Comment by bryancall** [comment]
[approve ci]

**Comment by atsci** [comment]
Linux build *failed*! https://ci.trafficserver.apache.org/job/linux-github/1713/

**Comment by atsci** [comment]
clang format *failed*! https://ci.trafficserver.apache.org/job/clang-format-github/126/

**Comment by atsci** [comment]
RAT check *failed*! https://ci.trafficserver.apache.org/job/RAT-github/140/

**Comment by atsci** [comment]
FreeBSD11 build *failed*! https://ci.trafficserver.apache.org/job/freebsd-github/1822/

**Comment by atsci** [comment]
clang-analyzer build *failed*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/384/

**Comment by atsci** [comment]
Intel CC build *failed*! https://ci.trafficserver.apache.org/job/icc-github/252/

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/123/

**Comment by bryancall** [comment]
[approve ci]

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/124/

**Comment by atsci** [comment]
Linux build *failed*! https://ci.trafficserver.apache.org/job/linux-github/1714/

**Comment by atsci** [comment]
clang format *failed*! https://ci.trafficserver.apache.org/job/clang-format-github/127/

**Comment by atsci** [comment]
RAT check *failed*! https://ci.trafficserver.apache.org/job/RAT-github/141/

**Comment by atsci** [comment]
FreeBSD11 build *failed*! https://ci.trafficserver.apache.org/job/freebsd-github/1823/

**Comment by atsci** [comment]
clang-analyzer build *failed*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/385/

**Comment by atsci** [comment]
Intel CC build *failed*! https://ci.trafficserver.apache.org/job/icc-github/253/

**Comment by bryancall** [comment]
[approve ci]

**Comment by atsci** [comment]
Linux build *failed*! https://ci.trafficserver.apache.org/job/linux-github/1715/

**Comment by atsci** [comment]
clang format *failed*! https://ci.trafficserver.apache.org/job/clang-format-github/128/

**Comment by atsci** [comment]
RAT check *failed*! https://ci.trafficserver.apache.org/job/RAT-github/142/

**Comment by atsci** [comment]
FreeBSD11 build *failed*! https://ci.trafficserver.apache.org/job/freebsd-github/1824/

**Comment by atsci** [comment]
clang-analyzer build *failed*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/386/

**Comment by atsci** [comment]
Intel CC build *failed*! https://ci.trafficserver.apache.org/job/icc-github/254/

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/125/

**Comment by bryancall** [comment]
[approve ci]

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/144/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/130/

**Comment by atsci** [comment]
AU check *successful*! https://ci.trafficserver.apache.org/job/autest-github/127/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/1717/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1826/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/256/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/388/

**Pull request merged**